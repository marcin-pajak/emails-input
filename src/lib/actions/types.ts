import { Elements, Handlers } from '../components/types';
import { Options } from '../options/types';
import { State, StateAction } from '../state/types';

/**
 * Email actions
 */
export interface EmailAddActions {
  add: (value: string | string[]) => boolean;
}

export interface EmailRemoveActions {
  removeByIndex: (index: number) => void;
  removeByLocalId: (localId: string) => void;
  removeByName: (name: string) => void;
}

export type EmailActions = EmailAddActions & EmailRemoveActions;

export interface EmailActionsProperties {
  elements: Elements;
  handleAction: (action: StateAction) => string | undefined;
  options: Options;
  state: State;
}

/**
 * Library actions
 */
export interface LibraryActions {
  destroy: () => void;
  init: () => void;
}

export interface LibraryActionProperties {
  elements: Elements;
  handleAction: (action: StateAction) => string | undefined;
  handlers: Handlers;
  state: State;
}

/**
 * Actions
 */
export type Actions = EmailActions & LibraryActions;
