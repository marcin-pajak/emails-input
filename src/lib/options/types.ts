import { ChangeType } from '../constants';

export interface Options {
  initialValue: string[];
  onChange: (changed: string, type: ChangeType, newValue: string[]) => void;
}
