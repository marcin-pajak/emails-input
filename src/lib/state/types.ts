import { ActionType, LibraryState } from './constants';

/**
 * State Actions
 */
export interface AddEmailAction {
  payload: string;
  type: typeof ActionType.Add;
}

export interface RemoveEmailAction {
  payload: string;
  type: typeof ActionType.Remove;
}

export interface UpdateStateAction {
  payload: LibraryState;
  type: typeof ActionType.UpdateState;
}

export type StateAction =
  | AddEmailAction
  | RemoveEmailAction
  | UpdateStateAction;

/**
 * State
 */
export interface Entry {
  email: string;
  isValid: boolean;
}

export type LocalId = string;

export type EmailMapper = Record<string, LocalId>;

export type EntryMap = Record<string, Entry>;

export interface State {
  emails: string[];
  entries: Entry[];
  entryMap: EntryMap;
  mapper: EmailMapper;
  state: LibraryState;
}

/**
 * State Manager
 */
export interface StateManager {
  handleAction: (action: StateAction, state: State) => string | undefined;
  makeState: (initialValue: string[]) => State;
}
