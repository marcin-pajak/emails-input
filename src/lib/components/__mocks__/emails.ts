export const emailsTemplate = (): string => '<div />';
