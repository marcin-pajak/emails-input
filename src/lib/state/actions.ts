import { isValidEmail } from '../../utils/isValidEmail';
import {
  AddEmailAction,
  RemoveEmailAction,
  State,
  UpdateStateAction,
} from './types';

/**
 * Remove email
 * @param action
 * @param state
 */
export const handleRemoveAction = (
  action: RemoveEmailAction,
  state: State,
): void => {
  const name = state.entryMap[action.payload].email;
  const index = state.emails.indexOf(name);
  delete state.entryMap[action.payload];
  delete state.mapper[name];
  state.entries.splice(index, 1);
  state.emails.splice(index, 1);
};

/**
 * Add email
 * Returns local Id of new email
 * @param action
 * @param state
 * @param newId
 */
export const handleAddAction = (
  action: AddEmailAction,
  state: State,
  newId: string,
): void => {
  const entry = {
    email: action.payload,
    isValid: isValidEmail(action.payload),
  };
  state.emails.push(action.payload);
  state.mapper[action.payload] = newId;
  state.entryMap[newId] = entry;
  state.entries.push(entry);
};

/**
 * Update Library state
 * @param action
 * @param state
 */
export const handleUpdateAction = (
  action: UpdateStateAction,
  state: State,
): void => {
  state.state = action.payload;
};
