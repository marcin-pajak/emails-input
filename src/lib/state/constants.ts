export enum ActionType {
  Add = 'ADD',
  Remove = 'REMOVE',
  Initialize = 'INITIALIZE',
  Destroy = 'DESTROY',
  UpdateState = 'UPDATE_STATE',
}

export enum LibraryState {
  Initializing = 'Initializing',
  Initialized = 'Initialized',
  Destroyed = 'Destroyed',
}
