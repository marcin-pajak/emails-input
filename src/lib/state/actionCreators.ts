import { ActionType, LibraryState } from './constants';
import { AddEmailAction, RemoveEmailAction, UpdateStateAction } from './types';

/**
 * Add new email
 * @param email
 */
export const addItemAction = (email: string): AddEmailAction => ({
  payload: email,
  type: ActionType.Add,
});

/**
 * Remove email
 * @param id
 */
export const removeItemAction = (id: string): RemoveEmailAction => ({
  payload: id,
  type: ActionType.Remove,
});

/**
 * Update Library's state
 * @param state
 */
export const updateStateAction = (state: LibraryState): UpdateStateAction => ({
  payload: state,
  type: ActionType.UpdateState,
});
