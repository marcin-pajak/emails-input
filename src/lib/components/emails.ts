import { objectEntries } from '../../utils/object';
import { Entry, EntryMap } from '../state/types';
import { emailTemplate } from './email';

/**
 * Emails Component Template
 * @param entryMap
 */
export const emailsTemplate = (entryMap: EntryMap): string =>
  objectEntries<Entry>(entryMap)
    .map(([localId, entry]: [string, Entry]): string =>
      emailTemplate(entry, localId),
    )
    .join('');
