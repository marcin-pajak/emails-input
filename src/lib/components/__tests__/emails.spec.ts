import { EntryMap } from '../../state/types';
import { emailsTemplate } from '../emails';

jest.mock('../email');

describe('emails template should', (): void => {
  it('return valid template', (): void => {
    expect.hasAssertions();

    const emailMap: EntryMap = {
      '1': {
        email: 'valid@email.com',
        isValid: true,
      },
      '2': {
        email: 'invalid',
        isValid: false,
      },
    };

    expect(emailsTemplate(emailMap)).toMatchSnapshot();
  });
});
