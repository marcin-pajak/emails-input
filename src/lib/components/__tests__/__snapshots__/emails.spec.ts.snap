// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emails template should return valid template 1`] = `"<span /><span />"`;
